java/code/src/com/redhat/rhn/manager/channel/ChannelManager.java

```diff
@@ -2573,14 +2573,14 @@ public static String getRepoLastBuild(Channel channel) {
     public static DataResult listPossibleSuseBaseChannelsForServer(Server s) {
         log.debug("listPossibleSuseBaseChannelsForServer called");
 
-        SUSEProductSet ps = s.getInstalledProductSet();
-        if (ps == null || ps.getBaseProduct() == null) {
+        Optional<SUSEProductSet> ps = s.getInstalledProductSet();
+        if (!ps.isPresent() || ps.get().getBaseProduct() == null) {
```

Comment:
fold on `Optional` does not work on Java 8, AFAIK. It would take an opt.map(...).orElseGet(...) instead, and I think that compromises readability on trivial situations like this.

Reply:
Suggested changes:

```
        Opt.fold(ps, 
                () -> {
                    log.info("Server has no base product installed");
                    return null; 
                },
                opt -> {
                    if (opt.getBaseProduct() == null) {
                        log.info("Server has no base product installed");
                        return null;
                    }
                    Map params = new HashMap();
                    params.put("pid", opt.getBaseProduct().getId());
                    params.put("channel_arch_id", s.getServerArch().getCompatibleChannelArch().getId());
                    SelectMode m3 = ModeFactory.getMode("Channel_queries",
                            "suse_base_channels_for_suse_product");
                    DataResult ret  = makeDataResult(params, new HashMap(), null, m3);
                    return ret;
                });
```
Caveat: Handling the case where opt.getBaseProduct() is null.

Suggestions?

Reply:
You don't really use `ps` or the `baseproduct` anywhere the only thing you need is the baseproducts id. A combination of `map` and `flatMap` should get you the `Optional` with the content you need and then fold gets exactly that.

Optimally tho this method would also just return `Optional` instead of `null` in which case you would not need the `fold` at all.

Reply:
Proposed solution:

```
    public static Optional<DataResult> listPossibleSuseBaseChannelsForServer(Server s) {
        log.debug("listPossibleSuseBaseChannelsForServer called");

        Optional<Long> baseProductId = s.getInstalledProductSet().flatMap(
                ps -> ofNullable(ps.getBaseProduct())).map(SUSEProduct::getId);

        return Opt.fold(baseProductId,
                () -> {
                    log.info("Server has no base product installed");
                    return empty();
                },
                id -> {
                    Map params = new HashMap();
                    params.put("pid", id);
                    params.put("channel_arch_id", s.getServerArch().getCompatibleChannelArch().getId());
                    SelectMode m3 = ModeFactory.getMode("Channel_queries",
                            "suse_base_channels_for_suse_product");
                    DataResult ret  = makeDataResult(params, new HashMap(), null, m3);
                    return of(ret);
                });
    }
``` 

Returning Optional on this method would sound like the best possibility, but seems there are some issues on Opt.fold in handling generic types. Changing the return type of the method above requires changes in a few places, being one of them on method `public static Optional<Channel> guessServerBaseChannel(User usr, Server s)` on the same class: 

```
List<EssentialChannelDto> dr = listPossibleSuseBaseChannelsForServer(s);
if (dr != null && !dr.isEmpty()) {
    c = ChannelFactory.lookupByIdAndUser(dr.get(0).getId(), usr);
}
```

has to be replaced with something like (in case we want to avoid using .get)

```
c = Opt.fold(listPossibleSuseBaseChannelsForServer(s),
        () -> null,
        dr -> ChannelFactory.lookupByIdAndUser(dr.get(0).getId(), usr));
```
However, `.getId()` is at this point an unresolvable method. Apart from modifying Opt.fold, we can either return null on `listPossibleSuseBaseChannelsForServer` or use .get here. What do you suggest?

Reply:
PR updated, please check